### PR TITLE
Improve request/progress performance

### PR DESCRIPTION
**Pull request by hjelmn**
This fixes a couple of performance regressions and is targeted to 2.0.0.

@nysal, @thananon Let me know if there is anything else that should be fixed. The RC needs to go out ASAP.

### CODE REVIEW

File: opal/threads/wait_sync.h [hunk 1]
```diff
@@ -82,12 +78,13 @@ OPAL_DECLSPEC int sync_wait_st(ompi_wait_sync_t *sync);
...
 {
     if( OPAL_LIKELY(OPAL_SUCCESS == status) ) {
-        if( 0 != (OPAL_ATOMIC_ADD_32(&sync->count, -updates)) ) {
+        if( 0 != (OPAL_THREAD_ADD32(&sync->count, -updates)) ) {
             return;
         }
     } else {
-        OPAL_ATOMIC_CMPSET_32(&(sync->count), 0, 0);
-        sync->status = -1;
+        /* this is an error path so just use the atomic */
```
Comment (hjelmn) [inline opal/threads/wait_sync.h]: Want to point out this error path was wrong. ```OPAL_ATOMIC_CMPSET_32(&(sync->count), 0, 0);``` does nothing.

**Comment by thananon** [comment]
👍 
Thank you, Nathan.

File: opal/threads/wait_sync.h [hunk 2]
```diff
@@ -27,50 +29,44 @@ typedef struct ompi_wait_sync_t {
...
-       (sync)->status = 0;                            \
-       PTHREAD_COND_INIT(&(sync)->condition, NULL);   \
-       PTHREAD_MUTEX_INIT(&(sync)->lock, NULL);       \
+#define WAIT_SYNC_INIT(sync,c)                                  \
+    do {                                                        \
+        (sync)->count = c;                                      \
+        (sync)->next = NULL;                                    \
+        (sync)->prev = NULL;                                    \
+        (sync)->status = 0;                                     \
+        if (opal_using_threads()) {                             \
```
Comment (nysal) [inline opal/threads/wait_sync.h]: Since PTHREAD_COND_INIT and PTHREAD_MUTEX_INIT already check for opal_using_threads(), this branch might be redundant ?

**Comment by nysal** [comment]
@hjelmn Apart from the previous comment it looks good :+1:

File: opal/threads/wait_sync.h (hunk 2 above)
Comment (hjelmn) [inline opal/threads/wait_sync.h]: Indeed. I didn't think to check what those macro did. Will clean it up in a bit.

Comment (hjelmn) [inline opal/threads/wait_sync.h]: Unless you want to do it. I can add your commit to the 2.x PR.

Comment (nysal) [inline opal/threads/wait_sync.h]: Didn't see your comment until later, timezone issues. I see that you already have a PR to fix it - https://github.com/open-mpi/ompi/pull/1811

Thanks :+1:

**Pull request merged**